glommio/src/executor/stall.rs

```diff
@@ -0,0 +1,305 @@
+}
+
+impl fmt::Debug for StallDetection<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "[stall-detector -- executor {}] task queue {} went over-budget: {:#?} (budget: \
+             {:#?}). Backtrace: {:#?}",
+            self.executor, self.task_queue, self.overage, self.budget, self.trace,
+        )
+    }
+}
+
+pub trait StallDetectionHandler: std::fmt::Debug + Send + Sync {
+    /// What signal number to use; see values in libc::SIG*
+    fn signal(&self) -> u8;
+
+    /// Handler called when a task exceeds its budget
+    fn stall(&self, detection: StallDetection<'_>);
+}
+
+#[derive(Debug)]
+pub struct DefaultStallDetectionHandler {}
+
+impl StallDetectionHandler for DefaultStallDetectionHandler {
+    fn signal(&self) -> u8 {
+        nix::libc::SIGUSR1 as u8
+    }
+
+    fn stall(&self, detection: StallDetection<'_>) {
+        log::warn!("{:?}", detection);
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct StallDetector {
+    timer: Arc<sys::timerfd::TimerFd>,
+    stall_handler: Box<dyn StallDetectionHandler + 'static>,
+    timer_handler: Option<JoinHandle<()>>,
+    id: usize,
+    terminated: Arc<AtomicBool>,
+    //NOTE: we don't use signal_hook::low_level::channel as backtraces
+    //      have too many elements
+    pub(crate) tx: crossbeam::channel::Sender<backtrace::BacktraceFrame>,
+    pub(crate) rx: crossbeam::channel::Receiver<backtrace::BacktraceFrame>,
+}
+
+impl StallDetector {
+    pub(crate) fn new(
+        executor_id: usize,
+        stall_handler: Box<dyn StallDetectionHandler + 'static>,
+    ) -> std::io::Result<StallDetector> {
+        let timer = Arc::new(
+            sys::timerfd::TimerFd::new(
+                sys::timerfd::ClockId::CLOCK_MONOTONIC,
+                sys::timerfd::TimerFlags::empty(),
+            )
+            .map_err(std::io::Error::from)?,
+        );
+        let tid = unsafe { nix::libc::pthread_self() };
+        let terminated = Arc::new(AtomicBool::new(false));
+        let sig = stall_handler.signal();
+        let timer_handler = std::thread::spawn(enclose::enclose! { (terminated, timer) move || {
+            while timer.wait().is_ok() {
+                if terminated.load(Ordering::Relaxed) {
+                    return
+                }
+                unsafe { nix::libc::pthread_kill(tid, sig.into()) };
+            }
+        }});
+        let (tx, rx) = crossbeam::channel::bounded(1 << 10);
+
+        Ok(Self {
+            timer,
+            timer_handler: Some(timer_handler),
+            stall_handler,
+            id: executor_id,
+            terminated,
+            tx,
+            rx,
+        })
+    }
+
+    pub(crate) fn enter_task_queue(
+        &self,
+        queue_name: String,
+        start: Instant,
+        max_expected_runtime: Duration,
+    ) -> nix::Result<StallDetectorGuard<'_>> {
+        // We consider a queue to be stalling the system if it failed to yield in due
```

Comment:
on the other hand for a 500us 10% does little if anything.

The application will know better what threshold to use, so I think it is best to have something user-defined than this heuristic.

Reply:
Agreed. @davidblewett let's use the handler object for that like we for deciding which signal to use.